include/ensmallen_bits/gradient_descent/gradient_descent_impl.hpp

```diff
@@ -26,31 +26,55 @@ inline GradientDescent::GradientDescent(
     const double tolerance) :
     stepSize(stepSize),
     maxIterations(maxIterations),
-    tolerance(tolerance)
+    tolerance(tolerance),
+    terminate(false)
 { /* Nothing to do. */ }
 
 //! Optimize the function (minimize).
-template<typename FunctionType>
-double GradientDescent::Optimize(
-    FunctionType& function, arma::mat& iterate)
+template<typename FunctionType,
+         typename MatType,
+         typename GradType,
+         typename... CallbackTypes>
+typename std::enable_if<IsArmaType<GradType>::value,
+typename MatType::elem_type>::type
+GradientDescent::Optimize(FunctionType& function,
+                          MatType& iterateIn,
+                          CallbackTypes&&... callbacks)
 {
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+  typedef typename MatTypeTraits<MatType>::BaseMatType BaseMatType;
+  typedef typename MatTypeTraits<GradType>::BaseMatType BaseGradType;
+
   // Use the Function<> wrapper type to provide additional functionality.
-  typedef Function<FunctionType> FullFunctionType;
+  typedef Function<FunctionType, BaseMatType, BaseGradType> FullFunctionType;
   FullFunctionType& f(static_cast<FullFunctionType&>(function));
 
   // Make sure we have the methods that we need.
-  traits::CheckFunctionTypeAPI<FullFunctionType>();
+  traits::CheckFunctionTypeAPI<FullFunctionType, BaseMatType, BaseGradType>();
+  RequireFloatingPointType<BaseMatType>();
+  RequireFloatingPointType<BaseGradType>();
+  RequireSameInternalTypes<BaseMatType, BaseGradType>();
 
   // To keep track of where we are and how things are going.
-  double overallObjective = std::numeric_limits<double>::max();
-  double lastObjective = std::numeric_limits<double>::max();
+  ElemType overallObjective = std::numeric_limits<ElemType>::max();
+  ElemType lastObjective = std::numeric_limits<ElemType>::max();
+
+  BaseMatType& iterate = (BaseMatType&) iterateIn;
 
   // Now iterate!
-  arma::mat gradient(iterate.n_rows, iterate.n_cols);
-  for (size_t i = 1; i != maxIterations; ++i)
+  terminate |= Callback::BeginOptimization(*this, f, iterate, callbacks...);
+  BaseGradType gradient(iterate.n_rows, iterate.n_cols);
+  for (size_t i = 1; i != maxIterations && !terminate; ++i)
   {
+    terminate |= Callback::BeginEpoch(*this, f, iterate, i, overallObjective,
+        callbacks...);
```

Comment:
Hmm, so it looks like we are using `BeginEpoch()` and `EndEpoch()` even when it the function is not separable.  This could be a little confusing, since people don't generally think the word `epoch` when they think about gradient descent (in fact, I suspect the term `epoch` is generally unknown outside of deep learning optimization).  I have a couple ideas:

(1) We could document clearly in the callback documentation that for non-separable functions, `BeginEpoch()` and `EndEpoch()` are called at the beginning and end of a single *iteration*.  (This works, since a non-separable optimizer is just like a separable optimizer except with `batchSIze = NumFunctions()`.)

(2) We could avoid the use of `BeginEpoch()` and `EndEpoch()` in non-separable optimizers, and instead have a `StepTaken()` callback only.  The reason I suggest `StepTaken()` and not `EvaluateWithGradient()` is that some non-separable optimizers (and some separable optimizers too) will propose multiple possible steps before actually taking one.  That can mean that `Evaluate()` or `EvaluateWithGradient()` is called potentially more than once per iteration.

Let me know what you think; maybe there are other options I didn't think of.

Reply:
For the second idea, do you still call the Evaluate/Gradient callback?

Reply:
I think we can leave the Evaluate/Gradient callbacks---they might still be useful if, e.g., you want to count the number of function evaluations or gradient evaluations, I guess.